FirebaseAnonymousAuthUI/FUIAnonymousAuth.m

```diff
@@ -0,0 +1,146 @@
+  FUIAuth *_authUI;
+
+  /** @var _presentingViewController
+      @brief The presenting view controller for interactive sign-in.
+   */
+  UIViewController *_presentingViewController;
+}
+
+- (instancetype)initWithAuthUI:(FUIAuth *)authUI {
+  if (self = [super init]) {
+    _authUI = authUI;
+  }
+  return self;
+}
+
+#pragma mark - FUIAuthProvider
+
+- (NSString *)providerID {
+  return nil;
+}
+
+/** @fn accessToken:
+    @brief Anonymous Auth token is matched by FirebaseUI User Access Token
+ */
+- (NSString *)accessToken {
+  return nil;
+}
+
+/** @fn idToken:
+    @brief Anonymous Auth Token Secret is matched by FirebaseUI User Id Token
+ */
+- (NSString *)idToken {
+  return nil;
+}
+
+- (NSString *)shortName {
+  return @"Anonymous";
+}
+
+- (NSString *)signInLabel {
+  return FUILocalizedStringFromTableInBundle(kSignInAsGuest, kTableName, kBundleName);
+}
+
+- (UIImage *)icon {
+  return [FUIAuthUtils imageNamed:@"ic_anonymous" fromBundle:kBundleName];
+}
+
+- (UIColor *)buttonBackgroundColor {
+  return [UIColor colorWithRed:244.0f/255.0f green:180.0f/255.0f blue:0.0f/255.0f alpha:1.0f];
+}
+
+- (UIColor *)buttonTextColor {
+  return [UIColor whiteColor];
+}
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-implementations"
+- (void)signInWithEmail:(nullable NSString *)email
+    presentingViewController:(nullable UIViewController *)presentingViewController
+                  completion:(nullable FIRAuthProviderSignInCompletionBlock)completion {
+  [self signInWithDefaultValue:email
+      presentingViewController:presentingViewController
+                    completion:completion];
+}
+#pragma clang diagnostic pop
+
+- (void)signInWithDefaultValue:(nullable NSString *)defaultValue
+      presentingViewController:(nullable UIViewController *)presentingViewController
+                    completion:(nullable FIRAuthProviderSignInCompletionBlock)completion {
+  [_authUI.auth signInAnonymouslyWithCompletion:^(FIRAuthDataResult * _Nullable authResult,
+                                                    NSError * _Nullable error) {
+    if (error) {
+      [FUIAuthBaseViewController showAlertWithMessage:error.localizedDescription
+                             presentingViewController:presentingViewController];
+      if (completion) {
+        completion(nil, error, nil);
+      }
+      return;
+    }
+    if (completion) {
+      completion(nil, error, nil);
+    }
+  }];
+}
+
+- (void)signOut {
+  FIRUser *user = _authUI.auth.currentUser;
+  __weak UIViewController *weakController = _presentingViewController;
+  if (user.isAnonymous) {
+    [user deleteWithCompletion:^(NSError * _Nullable error) {
+      if (error) {
+        __strong UIViewController *presentingViewController = weakController;
+        [FUIAuthBaseViewController showAlertWithMessage:error.localizedDescription
+                               presentingViewController:presentingViewController];
+        return;
+      }
+    }];
+  }
```

Comment:
that's something I'm not very sure about --- as far as I understand, we can't delete the user if it's not anonymous, otherwise a subsequent logout of other IDP (such as Google account) will fail.  In practice, if the user is upgrade from anonymous, it's no longer an anonymous user. logging out the other IDP is already sufficient to remove the user, and this should be a noop.